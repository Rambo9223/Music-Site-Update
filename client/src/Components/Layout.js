// page for links displayed in navbar
import { useState,useEffect } from "react";
import { Outlet,Link } from "react-router-dom"
import { Button } from "react-bootstrap";
import LogoutAdmin from "./Backend-Async/LogoutAdmin";
import AutoLogoutPopUp from "./AutoLogoutPopUp";
import Clock from "./Clock";

export default function Layout(){

  const [auth,setAuth] = useState(undefined);//admin jwt token
  const [response,setResponse] = useState();// server response variable 
  const [bool,setBool] = useState();// bool for popup display 
  const [send,setSend] = useState({type:"",message:""});//popup props
  // our local storage obejct user
  let stored = JSON.parse(localStorage.getItem("user"));

  /*The ToggleBool function is used 
  handle a auto logout and clear the local storage 
  and reload the page */

  function ToggleBool(){
    setBool(!bool);
    if(response!==null){
    if(response.ok===true){
        localStorage.clear();
        window.location.reload();
    }
}}

/*useEffect hook is used to check that there is a user
logged in and thus local storage variable stored has value */
  useEffect(()=>{
    if (stored!==null&&auth===undefined) {
        setAuth(stored.token);
    }
},[auth,stored])

  // layout depending on user
  let userLayout 
    // if admin
    if(auth!==undefined){
      userLayout = <ul className="nav-routes">
        
      <li>
        <Link className="route" to="/">Home</Link>
      </li>
      <li>
        <Link className="route" to="/news">News</Link>
      </li>
      <li>
        <Link className="route" to="/media">Media</Link>
      </li>
      <li>
        <Link className="route" to="/contact">Contact</Link>
      </li>
      <li>
        <Link className="route" to="/admin">Admin</Link>
      </li>
      <li>
      <div id="logout-button">
      <Button  variant="primary" className="rounded-pill" onClick={()=>{
        LogoutAdmin(auth).then((res)=>{// onclick logout admin passing token to server
          setResponse(res);// set response
          if(res.ok === true){//if successful 
            setBool(true);//set bool to true to bring up popup 
            // set popup information
            setSend({type:"Notification",
            message:`${res.message}`})
            }else{
            setBool(true);//set bool to true to bring up popup 
            setSend({// set information to error
            type:"Error",
            message:`Status - ${res.status}. ${res.message}`
            })
          }
        })
        
      } }>Logout</Button>
      </div>
      </li>
      <li id="clock">
        {/* The clock element works when admin is logged in and
        a valid auth token is present */}
          <Clock token={auth}/>
      </li>
      {/* if bool is true the popup element is show displaying a message to the user */}
      {bool===true?<AutoLogoutPopUp message={send.message} type={send.type} bool={bool} ToggleBool={ToggleBool}/>:null}
    </ul>
    }else{
      // user is not admin
      userLayout = <ul className="nav-routes">
      <li>
        <Link className="route" to="/">Home</Link>
      </li>
      <li>
        <Link className="route" to="/news">News</Link>
      </li>
      <li>
        <Link className="route" to="/media">Media</Link>
      </li>
      <li>
        <Link className="route" to="/contact">Contact</Link>
      </li>
      
    </ul>
    }
    return (
        <div>
             <nav className="route-container">
              {userLayout}
              </nav>
              <Outlet />
        </div>
    )
}