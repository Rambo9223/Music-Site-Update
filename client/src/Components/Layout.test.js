/**
*@jest-environment jsdom 
*/
import { render,cleanup} from '@testing-library/react';
import * as React from 'react';
import Layout from './Layout';
import { BrowserRouter as Router } from 'react-router-dom';

let token = "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJuYW1lIjoiR3Vlc3QiLCJ1c2VybmFtZSI6Ikd1ZXN0QWRtaW5AU1JNdXNpYy5jb20iLCJwYXNzd29yZCI6ImFkbWluMSIsImlhdCI6MTcyNTUzMDkxMSwiZXhwIjoxNzI1NTM0NTExfQ.dahD4ehVy35hz4Tv89jaqN9JpNy-3_-AOhxGEiRMohs"

afterEach(()=>{
  cleanup();
  localStorage.clear();  
});

function MockLayout(){
    return <Router><Layout/></Router>
}

describe('Layout Page Tests', () => {

    test('Page matches snapshot', () => {
        let page = render(<MockLayout/>);
        expect(page).toMatchSnapshot();
    });

    test('Page renders correct components without LocalStorage token', async () => {
        // in this case we expect with no token the admin nav link to not exist 

        let page = render(<MockLayout/>);
        let homePage = await page.findByText("Home");
        let adminPage = page.queryByText("Admin");
        expect(homePage).toBeInTheDocument();
        expect(adminPage).toEqual(null);
    });

    test('Page renders correct components with LocalStorage token', async () => {
        /* we create a local storage user token, now when page renders we expect to see the admin
        link element and logout button */
        const key = "user";
        const value = JSON.stringify({
            "token":token
        })
        localStorage.setItem(key,value);
        expect(localStorage.setItem).toHaveBeenLastCalledWith(key,value);
        expect(localStorage.__STORE__[key]).toBe(value);
        expect(Object.keys(localStorage.__STORE__).length).toBe(1);

        let page = render(<MockLayout/>);
        let adminPage = await page.findByText("Admin");
        let logoutButton = await page.findByText("Logout");
        expect(adminPage).toBeInTheDocument();
        expect(logoutButton).toBeInTheDocument();

    });
})