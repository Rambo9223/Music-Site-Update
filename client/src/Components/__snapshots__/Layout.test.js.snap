// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Layout Page Tests Page matches snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <nav
          class="route-container"
        >
          <ul
            class="nav-routes"
          >
            <li>
              <a
                class="route"
                href="/"
              >
                Home
              </a>
            </li>
            <li>
              <a
                class="route"
                href="/news"
              >
                News
              </a>
            </li>
            <li>
              <a
                class="route"
                href="/media"
              >
                Media
              </a>
            </li>
            <li>
              <a
                class="route"
                href="/contact"
              >
                Contact
              </a>
            </li>
          </ul>
        </nav>
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <nav
        class="route-container"
      >
        <ul
          class="nav-routes"
        >
          <li>
            <a
              class="route"
              href="/"
            >
              Home
            </a>
          </li>
          <li>
            <a
              class="route"
              href="/news"
            >
              News
            </a>
          </li>
          <li>
            <a
              class="route"
              href="/media"
            >
              Media
            </a>
          </li>
          <li>
            <a
              class="route"
              href="/contact"
            >
              Contact
            </a>
          </li>
        </ul>
      </nav>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
